# Assignment : Use iris dataset from seaborn and perform species prediction using GaussianNB(), MultinomialNB() & BernoulliNB() from Naive Bayes

## Loading the Libraries and the dataset

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns

In [2]:
k = sns.load_dataset('iris')
k.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [3]:
k.isna().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

In [4]:
k.species.value_counts()

setosa        50
versicolor    50
virginica     50
Name: species, dtype: int64

We will predict the 'species' (3 different species available) using 'sepal_length', 'sepal_width', 'petal_length' & 'petal_width' as inputs.  
Since the inputs are in float, we do not need to modify/typecast/encode the inputs  
Also there are no missing values

## Data Preparation

In [5]:
inp = k.iloc[:,:-1]
inp.head(2)

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2

In [6]:
out = k.species
out.head(2)

0    setosa
1    setosa
Name: species, dtype: object

**Importing the Naive_Bayes Libraries**

In [7]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB

## Prediction

**1) Prediction using GaussianNB()**

In [8]:
gnb = GaussianNB()
gnb.fit(inp,out)

GaussianNB()

In [9]:
k.sample(3)

sepal_length  sepal_width  petal_length  petal_width     species
69           5.6          2.5           3.9          1.1  versicolor
3            4.6          3.1           1.5          0.2      setosa
50           7.0          3.2           4.7          1.4  versicolor

In [10]:
gnb.predict([[5.9,3,5.2,0.7]])

array(['versicolor'], dtype='<U10')

In [11]:
gnb.predict_proba([[5.9,3,5.2,0.7]])

array([[3.45636199e-105, 9.97514922e-001, 2.48507837e-003]])

In [12]:
gnb.score(inp,out)*100

96.0

*Data Score for GaussianNB() is 96%*

**2) Prediction using MultinomialNB()**

In [13]:
mnb = MultinomialNB()
mnb.fit(inp,out)

MultinomialNB()

In [14]:
k.sample(3)

sepal_length  sepal_width  petal_length  petal_width    species
7             5.0          3.4           1.5          0.2     setosa
143           6.8          3.2           5.9          2.3  virginica
141           6.9          3.1           5.1          2.3  virginica

In [15]:
mnb.predict([[6.5,3.8,5.2,2.1]])

array(['virginica'], dtype='<U10')

In [16]:
k.species.value_counts()

setosa        50
versicolor    50
virginica     50
Name: species, dtype: int64

In [17]:
mnb.predict_proba([[6.5,3.8,5.2,2.1]])

array([[0.01943241, 0.4871864 , 0.49338119]])

In [18]:
mnb.score(inp,out)*100

95.33333333333334

*Data Score for GaussianNB() is 95.33%*

**3) Prediction using BernoulliNB()**

In [19]:
bnb = BernoulliNB()
bnb.fit(inp,out)

BernoulliNB()

In [20]:
k.sample(3)

sepal_length  sepal_width  petal_length  petal_width     species
35           5.0          3.2           1.2          0.2      setosa
85           6.0          3.4           4.5          1.6  versicolor
11           4.8          3.4           1.6          0.2      setosa

In [21]:
bnb.predict([[4.8,3.5,1.8,0.5]])

array(['setosa'], dtype='<U10')

In [22]:
k.species.value_counts()

setosa        50
versicolor    50
virginica     50
Name: species, dtype: int64

In [23]:
bnb.predict_proba([[4.8,3.5,1.8,0.5]])

array([[0.33333333, 0.33333333, 0.33333333]])

In [24]:
bnb.score(inp,out)*100

33.33333333333333

*Data Score for BernoulliNB() is 33.33%*

## Conclusion

We have predicted the Species feature in the iris dataset using its other features as input using GaussianNB(), MultinomialNB() & BernoulliNB() from the Naive Bayes Algorithm.
The data score has been highest for GaussianNB() at 96%, followed by MultinomialNB() at 95.33% & lowest for BernoulliNB() at 33.33%